internal/lsp/symbols.go

```diff
@@ -13,10 +13,39 @@ import (
 	"github.com/zclconf/go-cty/cty"
 )
 
+// defaultSymbols is the list of symbols that were supported by the initial
+// version of the LSP. This list is used as a fallback when the client does
+// not provide a list of supported symbols.
+var defaultSymbols = []lsp.SymbolKind{
+	lsp.File,
+	lsp.Module,
+	lsp.Namespace,
+	lsp.Package,
+	lsp.Class,
+	lsp.Method,
+	lsp.Property,
+	lsp.Field,
+	lsp.Constructor,
+	lsp.Enum,
+	lsp.Interface,
+	lsp.Function,
+	lsp.Variable,
+	lsp.Constant,
+	lsp.String,
+	lsp.Number,
+	lsp.Boolean,
+	lsp.Array,
+}
+
 func WorkspaceSymbols(sbs []decoder.Symbol, caps *lsp.WorkspaceSymbolClientCapabilities) []lsp.SymbolInformation {
 	symbols := make([]lsp.SymbolInformation, len(sbs))
+	supportedSymbols := defaultSymbols
+	if caps != nil && caps.SymbolKind != nil {
+		supportedSymbols = caps.SymbolKind.ValueSet
+	}
```

Comment:
Might be a bit far fetched, but does it maybe make sense to add a debug / trace log or something similar in an else block (here and below) to make it easier to spot if the default symbols fallback is used? I could see how that might aid in debugging why some (newer) capabilities are not activated 🤔
(Or are enabled capabilities already part of other log messages?)

Reply:
We log the capabilities as part of the `initialize` request and response.

<details><summary>Request for "initialize"</summary>
<p>

```json
{
	"processId": 66004,
	"clientInfo": {
		"name": "Visual Studio Code",
		"version": "1.86.0"
	},
	"locale": "en",
	"rootPath": "/Users/dbanck/tmp/funcs",
	"rootUri": "file:///Users/dbanck/tmp/funcs",
	"capabilities": {
		"workspace": {
			"applyEdit": true,
			"workspaceEdit": {
				"documentChanges": true,
				"resourceOperations": [
					"create",
					"rename",
					"delete"
				],
				"failureHandling": "textOnlyTransactional",
				"normalizesLineEndings": true,
				"changeAnnotationSupport": {
					"groupsOnLabel": true
				}
			},
			"configuration": true,
			"didChangeWatchedFiles": {
				"dynamicRegistration": true,
				"relativePatternSupport": true
			},
			"symbol": {
				"dynamicRegistration": true,
				"symbolKind": {
					"valueSet": [
						1,
						2,
						3,
						4,
						5,
						6,
						7,
						8,
						9,
						10,
						11,
						12,
						13,
						14,
						15,
						16,
						17,
						18,
						19,
						20,
						21,
						22,
						23,
						24,
						25,
						26
					]
				},
				"tagSupport": {
					"valueSet": [
						1
					]
				},
				"resolveSupport": {
					"properties": [
						"location.range"
					]
				}
			},
			"codeLens": {
				"refreshSupport": true
			},
			"executeCommand": {
				"dynamicRegistration": true
			},
			"didChangeConfiguration": {
				"dynamicRegistration": true
			},
			"workspaceFolders": true,
			"semanticTokens": {
				"refreshSupport": true
			},
			"fileOperations": {
				"dynamicRegistration": true,
				"didCreate": true,
				"didRename": true,
				"didDelete": true,
				"willCreate": true,
				"willRename": true,
				"willDelete": true
			},
			"inlineValue": {
				"refreshSupport": true
			},
			"inlayHint": {
				"refreshSupport": true
			},
			"diagnostics": {
				"refreshSupport": true
			}
		},
		"textDocument": {
			"publishDiagnostics": {
				"relatedInformation": true,
				"versionSupport": false,
				"tagSupport": {
					"valueSet": [
						1,
						2
					]
				},
				"codeDescriptionSupport": true,
				"dataSupport": true
			},
			"synchronization": {
				"dynamicRegistration": true,
				"willSave": true,
				"willSaveWaitUntil": true,
				"didSave": true
			},
			"completion": {
				"dynamicRegistration": true,
				"contextSupport": true,
				"completionItem": {
					"snippetSupport": true,
					"commitCharactersSupport": true,
					"documentationFormat": [
						"markdown",
						"plaintext"
					],
					"deprecatedSupport": true,
					"preselectSupport": true,
					"tagSupport": {
						"valueSet": [
							1
						]
					},
					"insertReplaceSupport": true,
					"resolveSupport": {
						"properties": [
							"documentation",
							"detail",
							"additionalTextEdits"
						]
					},
					"insertTextModeSupport": {
						"valueSet": [
							1,
							2
						]
					},
					"labelDetailsSupport": true
				},
				"insertTextMode": 2,
				"completionItemKind": {
					"valueSet": [
						1,
						2,
						3,
						4,
						5,
						6,
						7,
						8,
						9,
						10,
						11,
						12,
						13,
						14,
						15,
						16,
						17,
						18,
						19,
						20,
						21,
						22,
						23,
						24,
						25
					]
				},
				"completionList": {
					"itemDefaults": [
						"commitCharacters",
						"editRange",
						"insertTextFormat",
						"insertTextMode"
					]
				}
			},
			"hover": {
				"dynamicRegistration": true,
				"contentFormat": [
					"markdown",
					"plaintext"
				]
			},
			"signatureHelp": {
				"dynamicRegistration": true,
				"signatureInformation": {
					"documentationFormat": [
						"markdown",
						"plaintext"
					],
					"parameterInformation": {
						"labelOffsetSupport": true
					},
					"activeParameterSupport": true
				},
				"contextSupport": true
			},
			"definition": {
				"dynamicRegistration": true,
				"linkSupport": true
			},
			"references": {
				"dynamicRegistration": true
			},
			"documentHighlight": {
				"dynamicRegistration": true
			},
			"documentSymbol": {
				"dynamicRegistration": true,
				"symbolKind": {
					"valueSet": [
						1,
						2,
						3,
						4,
						5,
						6,
						7,
						8,
						9,
						10,
						11,
						12,
						13,
						14,
						15,
						16,
						17,
						18,
						19,
						20,
						21,
						22,
						23,
						24,
						25,
						26
					]
				},
				"hierarchicalDocumentSymbolSupport": true,
				"tagSupport": {
					"valueSet": [
						1
					]
				},
				"labelSupport": true
			},
			"codeAction": {
				"dynamicRegistration": true,
				"isPreferredSupport": true,
				"disabledSupport": true,
				"dataSupport": true,
				"resolveSupport": {
					"properties": [
						"edit"
					]
				},
				"codeActionLiteralSupport": {
					"codeActionKind": {
						"valueSet": [
							"",
							"quickfix",
							"refactor",
							"refactor.extract",
							"refactor.inline",
							"refactor.rewrite",
							"source",
							"source.organizeImports"
						]
					}
				},
				"honorsChangeAnnotations": false
			},
			"codeLens": {
				"dynamicRegistration": true
			},
			"formatting": {
				"dynamicRegistration": true
			},
			"rangeFormatting": {
				"dynamicRegistration": true
			},
			"onTypeFormatting": {
				"dynamicRegistration": true
			},
			"rename": {
				"dynamicRegistration": true,
				"prepareSupport": true,
				"prepareSupportDefaultBehavior": 1,
				"honorsChangeAnnotations": true
			},
			"documentLink": {
				"dynamicRegistration": true,
				"tooltipSupport": true
			},
			"typeDefinition": {
				"dynamicRegistration": true,
				"linkSupport": true
			},
			"implementation": {
				"dynamicRegistration": true,
				"linkSupport": true
			},
			"colorProvider": {
				"dynamicRegistration": true
			},
			"foldingRange": {
				"dynamicRegistration": true,
				"rangeLimit": 5000,
				"lineFoldingOnly": true,
				"foldingRangeKind": {
					"valueSet": [
						"comment",
						"imports",
						"region"
					]
				},
				"foldingRange": {
					"collapsedText": false
				}
			},
			"declaration": {
				"dynamicRegistration": true,
				"linkSupport": true
			},
			"selectionRange": {
				"dynamicRegistration": true
			},
			"callHierarchy": {
				"dynamicRegistration": true
			},
			"semanticTokens": {
				"dynamicRegistration": true,
				"tokenTypes": [
					"namespace",
					"type",
					"class",
					"enum",
					"interface",
					"struct",
					"typeParameter",
					"parameter",
					"variable",
					"property",
					"enumMember",
					"event",
					"function",
					"method",
					"macro",
					"keyword",
					"modifier",
					"comment",
					"string",
					"number",
					"regexp",
					"operator",
					"decorator",
					"hcl-attrName",
					"hcl-blockType",
					"hcl-blockLabel",
					"hcl-bool",
					"hcl-string",
					"hcl-number",
					"hcl-objectKey",
					"hcl-mapKey",
					"hcl-keyword",
					"hcl-referenceStep",
					"hcl-typeComplex",
					"hcl-typePrimitive",
					"hcl-functionName"
				],
				"tokenModifiers": [
					"declaration",
					"definition",
					"readonly",
					"static",
					"deprecated",
					"abstract",
					"async",
					"modification",
					"documentation",
					"defaultLibrary",
					"hcl-dependent",
					"terraform-data",
					"terraform-locals",
					"terraform-module",
					"terraform-output",
					"terraform-provider",
					"terraform-resource",
					"terraform-provisioner",
					"terraform-connection",
					"terraform-variable",
					"terraform-terraform",
					"terraform-backend",
					"terraform-name",
					"terraform-type",
					"terraform-requiredProviders"
				],
				"formats": [
					"relative"
				],
				"requests": {
					"range": true,
					"full": {
						"delta": true
					}
				},
				"multilineTokenSupport": false,
				"overlappingTokenSupport": false,
				"serverCancelSupport": true,
				"augmentsSyntaxTokens": true
			},
			"linkedEditingRange": {
				"dynamicRegistration": true
			},
			"typeHierarchy": {
				"dynamicRegistration": true
			},
			"inlineValue": {
				"dynamicRegistration": true
			},
			"inlayHint": {
				"dynamicRegistration": true,
				"resolveSupport": {
					"properties": [
						"tooltip",
						"textEdits",
						"label.tooltip",
						"label.location",
						"label.command"
					]
				}
			},
			"diagnostic": {
				"dynamicRegistration": true,
				"relatedDocumentSupport": false
			}
		},
		"window": {
			"showMessage": {
				"messageActionItem": {
					"additionalPropertiesSupport": true
				}
			},
			"showDocument": {
				"support": true
			},
			"workDoneProgress": true
		},
		"general": {
			"staleRequestSupport": {
				"cancel": true,
				"retryOnContentModified": [
					"textDocument/semanticTokens/full",
					"textDocument/semanticTokens/range",
					"textDocument/semanticTokens/full/delta"
				]
			},
			"regularExpressions": {
				"engine": "ECMAScript",
				"version": "ES2020"
			},
			"markdown": {
				"parser": "marked",
				"version": "1.1.0"
			},
			"positionEncodings": [
				"utf-16"
			]
		},
		"notebookDocument": {
			"synchronization": {
				"dynamicRegistration": true,
				"executionSummarySupport": true
			}
		},
		"experimental": {
			"refreshModuleProvidersCommandId": "client.refreshModuleProviders",
			"refreshModuleCallsCommandId": "client.refreshModuleCalls",
			"telemetryVersion": 1,
			"refreshTerraformVersionCommandId": "client.refreshTerraformVersion"
		}
	},
	"initializationOptions": {
		"validation": {
			"enableEnhancedValidation": true
		},
		"experimentalFeatures": {
			"validateOnSave": false,
			"prefillRequiredFields": true
		},
		"ignoreSingleFileWarning": false,
		"terraform": {
			"path": "",
			"timeout": "",
			"logFilePath": ""
		},
		"indexing": {
			"ignorePaths": [],
			"ignoreDirectoryNames": []
		}
	},
	"trace": "off",
	"workspaceFolders": [
		{
			"uri": "file:///Users/dbanck/tmp/funcs",
			"name": "funcs"
		}
	],
	"workDoneToken": "5603ed87-78ac-476f-8576-bb6ec589f775"
}
```

</p>
</details> 

<details><summary>Response to "initialize"</summary>
<p>

```json
{
	"capabilities": {
		"textDocumentSync": {
			"openClose": true,
			"change": 2,
			"save": {}
		},
		"completionProvider": {
			"triggerCharacters": [
				".",
				"["
			],
			"resolveProvider": true
		},
		"hoverProvider": true,
		"signatureHelpProvider": {
			"triggerCharacters": [
				"(",
				","
			]
		},
		"declarationProvider": true,
		"definitionProvider": true,
		"referencesProvider": true,
		"documentSymbolProvider": true,
		"codeActionProvider": {
			"codeActionKinds": [
				"source.formatAll.terraform"
			]
		},
		"codeLensProvider": {},
		"documentLinkProvider": {},
		"workspaceSymbolProvider": true,
		"documentFormattingProvider": true,
		"executeCommandProvider": {
			"commands": [
				"terraform-ls.module.callers",
				"terraform-ls.module.calls",
				"terraform-ls.module.providers",
				"terraform-ls.module.terraform",
				"terraform-ls.rootmodules",
				"terraform-ls.terraform.init",
				"terraform-ls.terraform.validate"
			],
			"workDoneProgress": true
		},
		"semanticTokensProvider": {
			"legend": {
				"tokenTypes": [
					"enumMember",
					"function",
					"keyword",
					"number",
					"parameter",
					"property",
					"string",
					"type",
					"variable",
					"hcl-attrName",
					"hcl-blockType",
					"hcl-blockLabel",
					"hcl-bool",
					"hcl-string",
					"hcl-number",
					"hcl-objectKey",
					"hcl-mapKey",
					"hcl-keyword",
					"hcl-referenceStep",
					"hcl-typeComplex",
					"hcl-typePrimitive",
					"hcl-functionName"
				],
				"tokenModifiers": [
					"defaultLibrary",
					"hcl-dependent",
					"terraform-backend",
					"terraform-connection",
					"terraform-data",
					"terraform-locals",
					"terraform-module",
					"terraform-name",
					"terraform-output",
					"terraform-provider",
					"terraform-provisioner",
					"terraform-requiredProviders",
					"terraform-resource",
					"terraform-terraform",
					"terraform-type",
					"terraform-variable"
				]
			},
			"full": true
		},
		"workspace": {
			"workspaceFolders": {
				"supported": true,
				"changeNotifications": "workspace/didChangeWorkspaceFolders"
			},
			"fileOperations": {}
		},
		"experimental": {
			"referenceCountCodeLens": false,
			"refreshModuleProviders": true,
			"refreshModuleCalls": true,
			"refreshTerraformVersion": true
		}
	},
	"serverInfo": {
		"name": "terraform-ls",
		"version": "0.33.0-dev"
	}
}
```

</p>
</details> 

Since falling back to the list of default symbols when none are sent is the behavior described in the spec and we don't have many issues regarding symbols, I would refrain from adding an extra log message to keep it clean.

We can always come back to this and extend it if something comes up.